enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
+* Allow users to configure the sizes that can be chosen for automatic scaling.
+* Allow users to configure the thresholds when scaling will occur for CPU and
+  memory usage.
+* Allow users to configure how often checks for automatic scaling should be run.
+* Allow users to configure how long after a scale up or scale down, no more
+  scaling should be performed.
+* Allow users to completely disable scale up.
+* Allow users to completely disable scale down.
+
+### Non-Goals
+
+* Horizontal scaling of control plane nodes
+
+## Proposal
+
+New configuration and a reconcile loop for the
+`control-plane-machine-set-operator` will be introduced that will monitor
+resource usage of the control plane nodes and automatically modify the
+`ControlPlaneMachineSet` if scaling the nodes is determined to be required.
+
+The new configuration will specify if scaling up and down is enabled and allow
+configuring when to trigger scaling. It will also allow configuring which
+machine types are to be used for scaling, putting an upper and lower limit on
+possible automatic scaling.
+
+### Workflow Description
+
+1. The OpenShift administrator installs the
+   `control-plane-machine-set-operator`.
+2. The OpenShift adminstrator creates a valid `control-plane-machine-set`
+   `configmap` in the `openshift-machine-api` namespace (or the respective
+   namespace `control-plane-machineset-operator` is running in), to configure
+   the automatic vertical scaling.
+
+### API Extensions
+
+This proposal does not require API extensions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation for this feature, requires a data source to gather current
+CPU and memory usage for nodes, as well as available resources on the nodes.
+
+To gather available resource information the default
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+to gather performance statistics for nodes. 
+
+Using the metrics API has the least requirements as it does not expect a running
+Prometheus monitoring stack. Using Prometheus generated data could make the
+implementation easier with respect to data aggregation.
+
+Data aggregation will be required using the metrics API, as decisions about
```

Comment:
I'd like to see an expansion on how this will work, sounds complicated

Reply:
I agree - this is essentially required if this is based solely on the metrics API.
If using Prometheus metrics it would be possible to simply retrieve the data from there instead of calculating it based on what is returned from the metrics API.
I will add an example, but IMO it might be better to use Prometheus as a data source to prevent this, but I can understand depending on the whole Prometheus stack (and granting access to it) comes with it's own complications.

Reply:
We have the potential to have the user configure the source right? We could have the metrics based implementation as the default with an option to switch over to the more efficient prometheus option via some config

Reply:
I have now reworked the whole threshold configuration to more closely match the configuration style of custom metrics autoscaler, including allowing to set a trigger to be prometheus based.